Nestjs init

jonseitz opened the pull request:
This turned into a bigger beast than I was initially expecting, as a lot of the tooling had to be ts-ified to get everything working in a mostly sane way.

At this point, I've only implemented the most basic setup with nest, keeping the default "hello world" endpoints from their starter, then adding `mongoose`, `cas-authentication`, and `webpack` middleware on top of that. I haven't added `typemoq` yet since there isn't much to actually test at this point. I'd appreciate any feedback you have on typescript best practices or other implementation details. 

At this point the server component of the project _should_ build correctly in docker (though not locally as the environment variables won't be loaded), though can expect to receive a long list of errors from webpack when it tries to compile the client code.

The next stage will be to move the client code over to Typescript. Once that's done I plan to add a very basic end-to-end implementation with testing.

jonseitz committed:
Update/install dependencies

Creates a tsconfig to setup typescript parser, Adds necessary
dependencies for typescript and nestjs, while also removing a
number of unneeded packages.

Also significantly streamlines the scripts section. May need to be
revisited after client code is setup.

jonseitz committed:
remove old server code

jonseitz committed:
add initial nestjs configs

jonseitz committed:
add dev middleware

jonseitz committed:
add cas-authentication middleware

jonseitz committed:
Replace jsdoc and apidoc with typedoc

jonseitz committed:
Set up mocha to run tests on typescript

jonseitz committed:
Create app user in mongo on start

Also makes a few tweaks to the default environment file.

jonseitz committed:
Replace babel with typescript

With typescript we no longer need the babel-loader to transpile ES-style
modules on the server, or to parse JSX. So I've cut out the babel
configurations in favor of ts-alternatives wherever possible.

Also swaps the babel-parser in eslint for the typescript parser, and
adds the typescript-eslint recommended rules, then tweaks a few things
that conflict.

jonseitz committed:
set up mongoose connection

jonseitz committed:
Configure mongodb session

Creates a session store in the database to track user data.

jonseitz committed:
Configure CAS authentication

Initializes the cas middleware in the main application

jonseitz committed:
Refactor constructor

Resolves eslint errors

jonseitz committed:
Add documentation blocks

Most documentation will be generated by typedoc based on type
declarations, but I added some short narrative comments to explain the
different components of the application.

rmainwork commented on webpackfile.js:
```diff
@@ -31,19 +34,22 @@ const client = {
     filename: 'server.js',
   },
+  resolve: {
+    extensions: ['.ts', '.js'],
```
I would maybe add `.tsx` in here in case we want to use JSX server side for email templates etc.

rmainwork replied:
Looks like they recommend that in the nest docs too: https://docs.nestjs.com/techniques/hot-reload#configuration

jonseitz replied:
Good point. added that in 821d248

rmainwork commented on src/server/middleware/cas.middleware.ts:
```diff
@@ -0,0 +1,69 @@
+        next();
+      }
+      if (!authorizationFilterAttribute) {
+        res.status(403).send('You do not have access to this application').end();
```
I would throw a [`ForbiddenException`](https://docs.nestjs.com/exception-filters#http-exceptions) here

```suggestion
      throw new UnauthorizedException('You do not have access to this application');
```

rmainwork replied:
Come to think of it, what about splitting this up into [a Guard](https://docs.nestjs.com/guards#authorization-guard) and middleware. Using the Guard for authorization and middleware for authentication?

jonseitz replied:
obviated by 5361903

rmainwork commented on src/server/middleware/cas.middleware.ts:
```diff
@@ -0,0 +1,69 @@
+        res.status(403).send('You do not have access to this application').end();
+        next();
+      }
+      res.status(401).send('This application requires HarvardKey').end();
```
I would throw an [`UnauthorizedException`](https://docs.nestjs.com/exception-filters#http-exceptions) here

```suggestion
      throw new UnauthorizedException('This application requires HarvardKey');
```

rmainwork replied:
See above comment regarding Guards

jonseitz replied:
obviated by 5361903

rmainwork commented on src/server/index.ts:
```diff
@@ -0,0 +1,15 @@
+import { NestFactory } from '@nestjs/core';
+import { AppModule } from './app.module';
+
+const { SERVER_PORT } = process.env;
```
Why not use [a configuration service](https://docs.nestjs.com/techniques/configuration)?

jonseitz replied:
Good idea. Set that up in a5e38da

rmainwork replied:
Did you end up using the config service here, or am I looking at old code?

rmainwork replied:
Actually, I just realized that wouldn't work because it's outside the app scope.

rmainwork commented on src/server/middleware/dev.middleware.ts:
```diff
@@ -0,0 +1,71 @@
+ * Configures react development for the client.
+ */
+
+const config = {
```
I wouldn't put webpack config with server middleware. Also, could some of this be re-used from the current webpack config?

jonseitz replied:
I believe this was settled by our separate conversation, but to recap: This webpack config is just to bundle and serve the client code while running the project in dev mode--part of the new hack-around to get simultaneously reloading code.

rmainwork replied:
Fair enough. Could any of the existing webpack config be imported and spread to be re-used though?

jonseitz replied:
It looks like the only top level fields in common are `name: 'client'` and `target: 'web'`, so I think importing and extending would probably be more work than it's worth

rmainwork replied:
Fair enough. Wasn't sure if any of the plugins etc. were shared.

rmainwork commented on src/server/middleware/session.middleware.ts:
```diff
@@ -0,0 +1,53 @@
+    authorizationFilterAttribute: boolean;
+  };
+}
+const MongoDBStore = connect(session);
```
Try to avoid connecting to mongo in middleware. I would suggest having this re-use the existing mongo connection, and then share the existing mongo connection around using a custom provider

jonseitz replied:
Very good point. Reworked that in dd6fdc7

rmainwork commented on src/server/middleware/cas.middleware.ts:
```diff
@@ -0,0 +1,69 @@
+import CASAuthentication from 'cas-authentication';
+import { SessionRequest } from './session.middleware';
+
+const {
```
See above regarding configuration service

jonseitz replied:
See a5e38da

rmainwork commented on src/server/middleware/cas.middleware.ts:
```diff
@@ -0,0 +1,69 @@
+} = process.env;
+
+
+export const cas = new CASAuthentication({
```
See above regarding configuration service

jonseitz replied:
See a5e38da

rmainwork commented on src/server/middleware/cas.middleware.ts:
```diff
@@ -0,0 +1,69 @@
+ */
+
+@Injectable()
+class CASMiddleware implements NestMiddleware {
```
This also might make more sense as a [Guard](https://docs.nestjs.com/guards)

jonseitz replied:
Yes it would. Created that guard in 5361903

rmainwork commented on src/server/index.ts:
```diff
@@ -0,0 +1,15 @@
+ * initializes and runs the nestjs app
+ */
+
+async function bootstrap(): Promise<void> {
```
Might be worth checking out [this](https://docs.nestjs.com/techniques/hot-reload) (if you haven't already)

jonseitz replied:
I think we also discussed this offline, but ever time I tried to create a double-hot-reloading situation (hot-reloading server code that hot-reloads client code) I always ran into a wall of errors. I'm hoping the current setup with `nodemon`/`ts-node` will work out for us, but we can revisit if the recompile time becomes an issue.

rmainwork commented on src/server/middleware/dev.middleware.ts:
```diff
@@ -0,0 +1,71 @@
+  },
+  plugins: [
+    new HtmlWebpackPlugin({
+      title: 'Docker App',
```
What about having this import the `name` from `package.json`?

jonseitz replied:
hmmm, I'd prefer not to use the `name` field as that's really the name for the whole project, more than the client app. But I did add an environment variable for this in 6ee2e7b (that way we can also change the displayed name in QA)

rmainwork replied:
That's a good idea!

rmainwork commented on src/server/middleware/session.middleware.ts:
```diff
@@ -0,0 +1,53 @@
+  },
+  collection: 'sessions',
+  databaseName: DB_DATABASE,
+}, (error: Error): void => { console.error(error); });
```
This would maybe be a good opportunity to use [the NestJS Logger](https://docs.nestjs.com/techniques/logger) to connect to CloudWatch, console.log etc.

jonseitz replied:
This is a good point, though at this point I don't believe it's 100% set that we'll be using cloudwatch for all logging. Let's take a look at some of the other logging libraries out there to see what makes the most sense for our needs and address this separately.

rmainwork replied:
Sure, but using a logging service should provide us with that flexibility - so we can swap out the underlying implementation as we see fit (perhaps we could use the built-in nest one to start with and then write our own for cloudwatch later?)

rmainwork commented on src/server/middleware/session.middleware.ts:
```diff
@@ -0,0 +1,53 @@
+ * and used to retrieve the values.
+ */
+
+export const sessionInit = session({
```
Just to link this back to the conversation we just had regarding CAS/passport - this is what I was looking at that sparked my thought...

rmainwork commented on mongoInit.js:
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable */
+db.runCommand({
```
I'd suggest that a user entity and repository be created and then this call those to create the necessary records(reduce duplication, type safety etc. etc.).

rmainwork replied:
Just had a thought here. I think you can get access to the nest application context outside of nest [like so](https://docs.nestjs.com/application-context) and from there you might be able to grab an instance of a service and create your user that way?

This would de-couple the script from any specific database implementation.

jonseitz replied:
Since this is very specifically [a script to be executed in the mongo shell](https://docs.mongodb.com/manual/tutorial/write-scripts-for-the-mongo-shell/) it really can't be decoupled from the mongo implementation (AFAIK you can't even `import` or `require` external libraries within the mongo shell).

As it happens, I converted it to a shell script in 4cf25b4 -- which allows us to actually use the container's environment variables. That should make it much easier to get set up.

rmainwork replied:
I was more thinking of calling the actual app via CLI and therefore re-using any database and entity stuff contained within it rather than having additional scripts to call and maintain. Not sure what you think of that as an idea....

rmainwork replied:
Just to recap on this for future reference - we decided we can't do this because the external script is required to provision the access and DB for the app so that the app can connect in the first place....

rmainwork reviewed (changes requested):
Great start overall, a few things I'd suggest tweaking or changing. I'm happy to help out with that if need be. :)

jonseitz committed:
add tsx to server extensions

jonseitz committed:
Convert mongo init script to sh

Swtiched from javascript to bash for mongo init to use environment vars

jonseitz committed:
Create ConfigModule

Manages environment variables within a global module

jonseitz committed:
Initialize session in middleware

jonseitz committed:
Implement HarvardKey Authentication

Sets up SAML authentication against Harvard Key to fetch user data (and
eventually permissions via Grouper). Also adds an AuthGuard based on the
default @nestjs/passport guard -- only difference is that it will
allow all traffic through while running in dev mode.

jonseitz committed:
Fully implement base nest module

jonseitz committed:
Fix MongoModule initialization

The MongoModule was incorrectly set up originally, so I changed the
getter on the config module to return the appropriate format. This
results in a simpler implementation, without the need for a separate
interface for the connection details

jonseitz committed:
Add APP_NAME to env

Sets a name for the application to be used as the title of the
index.html. Allows for customized titles on QA/Dev/Prod.

rmainwork commented on src/server/services/saml.strategy.ts:
```diff
@@ -0,0 +1,47 @@
+      };
+    }
+    if (!profile) {
+      throw new UnauthorizedException();
```
I would provide some kind of message here that we can display on the client side that provides more context...

```suggestion
      throw new UnauthorizedException('You are not authorized to use this application. Please contact SEAS computing');
```

That should give us something like the following:

```json
{
  "statusCode": 401,
  "error": "Unauthorized",
  "message": "You are not authorized to use this application. Please contact SEAS computing"
}
```

rmainwork commented on src/server/services/saml.strategy.ts:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  public async validate(profile: HarvardKeyProfile): Promise<UserData> {
+    if (this.devMode) {
```
I wonder about extracting this out into a separate instance that always returns true that we can use in dev mode.

rmainwork replied:
Opted not to do this since it would just move the `if this.devMode` to a custom provider without providing any real visible benefit

rmainwork commented on src/server/services/saml.strategy.ts:
```diff
@@ -0,0 +1,47 @@
+
+  public async validate(profile: HarvardKeyProfile): Promise<UserData> {
+    if (this.devMode) {
+      return {
```
I would recommend turning `UserData` into a class (essentially a DTO) and then returning an instance of that instead. This way, we don't have different data formats floating around (hopefully). Maybe create a folder under `src/services` and keep all the harvardkey stuff together?

jonseitz replied:
Added the user class in ac76a66

Is that what you had in mind?

rmainwork commented on src/server/services/saml.strategy.ts:
```diff
@@ -0,0 +1,47 @@
+    if (!profile) {
+      throw new UnauthorizedException();
+    }
+    return {
```
See [above comment regarding DTO](https://github.com/seas-computing/react-boilerplate/pull/5/files#r286526078)

jonseitz replied:
see ac76a66

rmainwork commented on src/server/decorators/auth.guard.ts:
```diff
@@ -0,0 +1,33 @@
+import { ConfigService } from '../services';
+
+/**
+ * Extends Passport's default AuthGuard with a new implementation
```
I wonder if this also should have a dummy version that just returns `true` in dev? What do you think?

rmainwork replied:
Opted not to do this since it would just move the `if this.devMode` to a custom provider without providing any real visible benefit(as above)

rmainwork commented on src/server/middleware/dev.middleware.ts:
```diff
@@ -0,0 +1,71 @@
+  plugins: [
+    new HtmlWebpackPlugin({
+      title: process.env.APP_NAME,
+      template: resolve(__dirname, '.src/client/index.html'),
```
To resolve the path issues you were having here - what do you think of exporting `SRC` from webpackfile.js? And then importing it here?

jonseitz replied:
e355ee0 adds another webpack plugin that injects the `<div id="root">` so we no longer need the html template.

rmainwork commented on src/server/services/config.service.ts:
```diff
@@ -0,0 +1,59 @@
+ * Parses process.env to create a clean configuration interface
+ */
+
+class ConfigService {
```
You've probably already seen this - but what do you think of using [config validation](https://docs.nestjs.com/techniques/configuration#validation)?

jonseitz replied:
Per our offline discussion, we'll consider this as a future enhancement but not as part of the pr.

jonseitz committed:
Provide message with error

Co-Authored-By: Robert Main <50675045+rmainseas@users.noreply.github.com>

jonseitz committed:
Create User Class

Adds a basic user class for use with the SAMl
authentication/authorization

jonseitz committed:
Use webpack root element plugin

Adds an additional webpack plugin that hooks into the existing
html-webpack-plugin to add <div id="root" /> to the body of the
generated index.html so we no longer need to specify the path to the
template file.

In the process of adding this I found a weird typescript quirk where
defining the configuration in a variable beforehand resulted in a type
error on the "mode" value. Upshot is that the passing the configuration
directly into the webpack() function resolves it.

rmainwork commented on src/server/middleware/dev.middleware.ts:
```diff
@@ -49,13 +50,13 @@ const config = {
+}]);
 
-const compiler = webpack(config);
+// const compiler = webpack([config]);
```
Might want to remove this commented out code

jonseitz replied:
d'oh. Dropped it in 8400038

jonseitz committed:
remove old code
